packages/exp/src/commands/upload/BaseUploader.js

```diff
@@ -0,0 +1,249 @@
+    const SOURCES_LIST = [
+      {
+        name: 'Latest build',
+        value: 'latest',
+      },
+      {
+        name: 'Other build',
+        value: 'other',
+      },
+    ];
+    const { source } = await prompt({
+      name: 'source',
+      message: 'Which build do you want to upload?',
+      type: 'list',
+      choices: SOURCES_LIST,
+    });
+
+    if (source === 'latest') return this.latestGetter();
+    else return this.chooseBuild();
+  };
+
+  getSource = () => {
+    const options = this.options;
+    let getter;
+    if (options.path) getter = this.pathGetter;
+    else if (options.latest) getter = this.latestGetter;
+    else if (options.id) getter = this.idGetter;
+    else getter = this.nothingGetter;
+
+    return {
+      getData: getter,
+    };
+  };
+
+  getConfigData = async () => {
+    const { exp } = await ProjectUtils.readConfigJsonAsync(this.projectDir);
+    const configName = await ProjectUtils.configFilenameAsync(this.projectDir);
+    if (!exp) {
+      throw new Error(`Couldn't read ${configName} file in project at ${this.projectDir}`);
+    }
+
+    return {
+      name: exp.name,
+      android: exp.android,
+      ios: exp.ios,
+      version: exp.version,
+      debug: { configName },
+    };
+  };
+
+  copyFile = async file => {
+    if (!fs.existsSync(file)) {
+      throw new Error(`File ${file} doesn't exist`);
+    }
+    const suffix = this.getPlatform() === 'ios' ? 'ipa' : 'apk';
+    const dest = `./build.${suffix}`;
+    await fs.copyFileSync(file, dest);
+    return dest;
+  };
+
+  downloadFileFrom = async remoteUrl => {
+    if (!/^http/.test(remoteUrl)) {
+      return this.copyFile(remoteUrl);
+    }
+    const dest = _.last(remoteUrl.split('/'));
+    if (fs.existsSync(dest)) {
+      log.warn(
+        `File ${dest} exists. If it's not ${this.getPlatformExtension()} you want to upload change its name`
+      );
+      return dest;
+    }
+    log(`Downloading build from ${remoteUrl}`);
+    const response = await Axios({
+      method: 'GET',
+      url: remoteUrl,
+      responseType: 'stream',
+    });
+    const totalLength = parseInt(response.headers['content-length'], 10);
+    const bar = new ProgressBar('Downloading [:bar] :percent :etas', {
+      complete: '=',
+      incomplete: ' ',
+      total: totalLength,
+    });
+    response.data.pipe(fs.createWriteStream(dest));
+    return new Promise((resolve, reject) => {
+      response.data.on('end', () => resolve(dest));
+      response.data.on('data', data => bar.tick(data.length));
+      response.data.on('error', error => reject(new Error(`${error}`)));
+    });
+  };
+
+  upload = async () => {
+    this.ensureOptionsAreCorrect();
+    const source = this.getSource();
+    const data = await source.getData();
+    const { remoteUrl } = data;
+    const configData = await this.getConfigData();
+    this.ensureConfigDataIsCorrect(configData);
+    const path = data.path ? data.path : await this.downloadFileFrom(remoteUrl);
```

Comment:
You've destructured `remoteUrl` from `data`, but didn't do the same with `path`, why?

Reply:
`downloadFileFrom` doesn't seem to be a good name, why not just `downloadFile`?